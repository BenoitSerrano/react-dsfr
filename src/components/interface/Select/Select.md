### Liste déroulante

```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
  />
```

### État d'erreur
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    message="Texte d'erreur obligatoire"
    messageType="error"
  />
```
### État de succès
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    message="Texte de validation"
    messageType="valid"
  />
```
### État inactif
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    disabled={true}
  />
```
### Liste déroulante avec texte d’aide
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    hint="Texte de description additionnel"
    options={options}
  />
```
