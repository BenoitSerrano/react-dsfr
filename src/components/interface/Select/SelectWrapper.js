import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';
import dataAttributes from '../../../utils/data-attributes';

import '@gouvfr/dsfr/dist/css/selects.min.css';

/**
 *
 * @visibleName SelectWrapper
 */
const SelectWrapper = ({
  className,
  hint,
  selectId,
  label,
  message,
  messageType,
  required,
  children,
  ...remainingProps
}) => {
  const [messageId] = useState(uuidv4());
  const _classNameWrapper = classNames('fr-select-group', {
    [`fr-select-group--${messageType}`]: messageType,
  }, className);

  return (
    <div
      className={_classNameWrapper}
      {...dataAttributes(remainingProps)}
    >
      {
      label && (
        <label
          className="fr-label"
          htmlFor={selectId}
          aria-describedby={messageId}
        >
          {label}
          {required && <span className="error"> *</span>}
          {hint && <span className="fr-hint-text" id={`${selectId}-desc-hint`}>{hint}</span>}
        </label>
      )
      }
      {children}
      {(message && messageType) && <p id={messageId} className={`fr-${messageType}-text`}>{message}</p>}
    </div>
  );
};

SelectWrapper.defaultProps = {
  className: '',
  hint: '',
  label: '',
  message: '',
  messageType: undefined,
  required: false,
};

SelectWrapper.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  hint: PropTypes.string,
  selectId: PropTypes.string.isRequired,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  required: PropTypes.bool,
  children: PropTypes.node.isRequired,
};

export default SelectWrapper;
