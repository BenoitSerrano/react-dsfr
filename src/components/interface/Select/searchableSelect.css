.select-search-options {
    background-color: var(--w);
    box-shadow: 0px 0px 10px var(--g300);
    position: absolute;
    z-index: 10;
    width: 100%;
    max-height: 32vh;
    overflow: scroll;
    display: none;
}

.select-search-options__visible {
    display: inherit;
}

.select-search-option__disabled,
.select-search-option {
    padding: 8px 16px;
    color: var(--bf500);
}

.select-search-option {
    cursor: pointer;
}

.select-search-option:hover {
    background-color: var(--block-color-hover);
}
