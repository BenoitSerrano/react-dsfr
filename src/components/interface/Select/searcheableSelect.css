.select-search-options {
    background-color: white;
    box-shadow: 0px 4px 30px rgba(43, 43, 43, 0.13);
    position: absolute;
    z-index: 10;
    width: 100%;
    max-height: 32vh;
    overflow: scroll;
    display: none;
}

.select-search-options__visible {
    display: inherit;
}

.select-search-option__disabled,
.select-search-option {
    padding: 8px 16px;
    color: var(--bf500);
}

.select-search-option {
    cursor: pointer;
}

.select-search-option:hover {
    background-color: var(--block-color-hover);
}
