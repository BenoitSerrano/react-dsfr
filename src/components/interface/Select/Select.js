import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';

/**
 *
 * @visibleName Select
 */
const Select = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
  required,
  ...remainingProps
}) => {
  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });
  const selectId = useRef(id || uuidv4());

  useEffect(() => {
    selectId.current = id || uuidv4();
  }, [id]);

  return (
    <SelectWrapper
      className={className}
      hint={hint}
      selectId={selectId.current}
      label={label}
      message={message}
      messageType={messageType}
      required={required}
      {...remainingProps}
    >
      <select
        className={_className}
        disabled={disabled}
        id={selectId.current}
        name={name}
        onChange={onChange}
        value={selected}
        data-testid="select"
      >
        {
          options.map((opt) => (
            <option
              disabled={opt.disabled || false}
              hidden={opt.hidden || false}
              key={`${selectId.current}-${opt.value}`}
              value={opt.value}
            >
              {opt.label}
            </option>
          ))
        }
      </select>
    </SelectWrapper>
  );
};

Select.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  selected: '',
  required: false,
};

Select.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  required: PropTypes.bool,
};

export default Select;
