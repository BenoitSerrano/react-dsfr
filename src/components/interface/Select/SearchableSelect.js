import React, {
  createRef, useEffect, useRef, useState,
} from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';
import './searchableSelect.css';

/**
 *
 * @visibleName SearchableSelect
 */
const SearchableSelect = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
  required,
  filter,
  ...remainingProps
}) => {
  const selectId = useRef(id || uuidv4());
  const optionsRef = useRef([]);
  const optionsContainerRef = useRef();
  const [arrowSelected, setArrowSelected] = useState();

  const getSelectedLabel = () => {
    if (selected) {
      const selectedOption = options.find((option) => option.value === selected);
      return selectedOption ? selectedOption.label : '';
    }
    return '';
  };

  const [internalLabel, setInternalLabel] = useState(getSelectedLabel());
  const [showOptions, setShowOptions] = useState(false);

  if (options.length !== optionsRef.length) {
    optionsRef.current = Array(options.length)
      .fill()
      .map((option, i) => optionsRef.current[i] || createRef());
  }

  useEffect(() => {
    selectId.current = id || uuidv4();
  }, [id]);

  useEffect(() => {
    if (arrowSelected) {
      optionsContainerRef.current.scrollTop = Math.max(
        0,
        optionsRef.current[arrowSelected].current.offsetTop - 20,
      );
    } else {
      optionsContainerRef.current.scrollTop = 0;
    }
  }, [arrowSelected]);

  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });

  const filteredOptions = options
    .filter((option, index, arr) => filter(internalLabel, option, index, arr));

  const onInternalChange = (newValue, newLabel) => {
    setInternalLabel(newLabel);
    onChange(newValue);
  };

  const onFocus = () => {
    onInternalChange('', '');
    setShowOptions(true);
    setArrowSelected(null);
  };

  const onBlur = () => {
    if (filteredOptions.length === 1) {
      onInternalChange(filteredOptions[0].value, filteredOptions[0].label);
    } else {
      const foundValue = options.find((option) => option.label === internalLabel);
      if (!foundValue) {
        setInternalLabel('');
      }
    }
    setShowOptions(false);
  };

  const onKeyDown = (e) => {
    switch (e.key) {
      case 'ArrowDown':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected === null) {
          setArrowSelected(0);
        } else if (arrowSelected < filteredOptions.filter((o) => !o.disabled).length - 1) {
          setArrowSelected(arrowSelected + 1);
        }
        break;
      case 'ArrowUp':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected && arrowSelected > 0) {
          setArrowSelected(arrowSelected - 1);
        }
        break;
      case 'Enter':
        e.preventDefault();
        if (arrowSelected !== null) {
          const option = filteredOptions.filter((o) => !o.disabled)[arrowSelected];
          onInternalChange(option.value, option.label);
          setShowOptions(false);
        }
        break;
      default:
        setArrowSelected(null);
    }
  };

  let refCount = -1;
  return (
    <SelectWrapper
      className={className}
      hint={hint}
      selectId={selectId.current}
      label={label}
      message={message}
      messageType={messageType}
      required={required}
      {...remainingProps}
    >
      <div className="select-search-input">
        <input
          onKeyDown={onKeyDown}
          id={selectId.current}
          className={_className}
          autoComplete="off"
          required={required}
          disabled={disabled}
          onChange={(e) => setInternalLabel(e.target.value)}
          onFocus={onFocus}
          onBlur={onBlur}
          value={internalLabel}
        />
        <div
          ref={optionsContainerRef}
          className={classNames(
            'select-search-options',
            'midlength-input',
            { 'select-search-options__visible': showOptions },
          )}
        >
          {filteredOptions.length === 0 ? (
            <div className="select-search-option__disabled">
              Aucun r√©sultat
            </div>
          ) : (
            <>
              {filteredOptions.map((option) => {
                if (!option.disabled) {
                  refCount += 1;
                }
                return (
                  <option
                    ref={option.disabled ? null : optionsRef.current[refCount]}
                    className={classNames(
                      'select-search-option',
                      {
                        'select-search-option__selected': !option.disabled && refCount === arrowSelected,
                        'select-search-option__disabled': option.disabled,
                      },
                    )}
                    disabled={option.disabled || false}
                    hidden={option.hidden || false}
                    key={`${selectId}-${option.value}`}
                    value={option.value}
                    onMouseDown={() => onInternalChange(option.value, option.label)}
                  >
                    {option.label}
                  </option>
                );
              })}
            </>
          )}
        </div>
      </div>
    </SelectWrapper>
  );
};

SearchableSelect.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  selected: '',
  required: false,
  filter: (
    label,
    option,
  ) => option.label.toLowerCase().includes(label.toLowerCase()),
};

SearchableSelect.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.string,
  required: PropTypes.bool,
  filter: PropTypes.func,
};

export default SearchableSelect;
