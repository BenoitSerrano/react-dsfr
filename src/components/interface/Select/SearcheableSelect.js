import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';
import './searcheableSelect.css';

/**
 *
 * @visibleName SearcheableSelect
 */
const SearcheableSelect = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
  required,
  filter,
  ...remainingProps
}) => {
  const [selectId] = useState(id || uuidv4());

  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });

  const [internalValue, setInternalValue] = useState(selected);
  const [showOptions, setShowOptions] = useState(false);

  const filteredOptions = options
    .filter((option, index, arr) => filter(internalValue, option, index, arr));

  const onInternalChange = (value, newValue) => {
    setInternalValue(newValue);
    onChange(value);
  };

  const onFocus = () => {
    onInternalChange('', '');
    setShowOptions(true);
  };

  const onBlur = () => {
    if (filteredOptions.length === 1) {
      onInternalChange(filteredOptions[0].value, filteredOptions[0].label);
    } else {
      const foundValue = options.find((option) => option.label === internalValue);
      if (!foundValue) {
        setInternalValue('');
      }
    }
    setShowOptions(false);
  };

  return (
    <SelectWrapper
      className={className}
      hint={hint}
      selectId={selectId}
      label={label}
      message={message}
      messageType={messageType}
      required={required}
      {...remainingProps}
    >
      <input
        id={selectId}
        className={_className}
        autoComplete="off"
        required={required}
        disabled={disabled}
        onChange={(e) => setInternalValue(e.target.value)}
        onFocus={onFocus}
        onBlur={onBlur}
        value={internalValue}
      />
      <div
        className={classNames(
          'select-search-options',
          'midlength-input',
          { 'select-search-options__visible': showOptions },
        )}
      >
        {filteredOptions.length === 0 ? (
          <div className="select-search-option__disabled">
            Aucun r√©sultat
          </div>
        ) : (
          <>
            {filteredOptions.map((option) => (
              <option
                className="select-search-option"
                disabled={option.disabled || false}
                hidden={option.hidden || false}
                key={`${selectId}-${option.value}`}
                value={option.value}
                onMouseDown={() => onInternalChange(option.value, option.label)}
              >
                {option.label}
              </option>
            ))}
          </>
        )}
      </div>
    </SelectWrapper>
  );
};

SearcheableSelect.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  selected: '',
  required: false,
  filter: (
    internalValue,
    option,
  ) => option.label.toLowerCase().includes(internalValue.toLowerCase()),
};

SearcheableSelect.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.string,
  required: PropTypes.bool,
  filter: PropTypes.func,
};

export default SearcheableSelect;
