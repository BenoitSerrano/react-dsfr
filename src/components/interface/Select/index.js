import SearcheableSelect from './SearcheableSelect';
import Select from './Select';

export {
  SearcheableSelect,
  Select,
};
