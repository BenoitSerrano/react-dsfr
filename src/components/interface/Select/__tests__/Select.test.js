import { render, screen } from '@testing-library/react';
import Select from '..';

describe('<Select />', () => {
  const options = [
    { value: '1', label: 'label-1' },
    { value: '2', label: 'label-2' },
  ];

  it('in document', () => {
    render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    const search = screen.getByTestId('select');
    expect(search).toBeInTheDocument();
  });

  // test de la présence de la class rf-select
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    expect(container.getElementsByClassName('rf-select').length).toBe(1);
  });

  // test de la class d'erreur
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="error"
      />,
    );
    expect(container.getElementsByClassName('rf-select--error').length).toBe(1);
  });

  // test de la class succès
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    expect(container.getElementsByClassName('rf-select--valid').length).toBe(1);
  });
});
