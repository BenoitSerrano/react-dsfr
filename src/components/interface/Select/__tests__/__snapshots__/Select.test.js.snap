// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders correctly 1`] = `
<div
  className="rf-select-group rf-select-group--valid"
>
  <label
    aria-describedby="16f18b04-0439-4d91-be64-f4c12eaf7e35"
    className="rf-label"
    htmlFor="b74a1772-2811-4d88-ae43-87a192704043"
  >
    Label pour liste d√©roulante
    
  </label>
  <select
    className="rf-select rf-select--valid"
    data-testid="select"
    disabled={false}
    id="b74a1772-2811-4d88-ae43-87a192704043"
    name={null}
    onChange={[Function]}
    value=""
  >
    <option
      disabled={false}
      hidden={false}
      value="1"
    >
      label-1
    </option>
    <option
      disabled={false}
      hidden={false}
      value="2"
    >
      label-2
    </option>
  </select>
  
</div>
`;
